brainiak/funcalign/rsrm.py

```diff
@@ -0,0 +1,560 @@
+            logger.info('Objective function %f' % objective)
+
+        # Main loop
+        for i in range(self.n_iter):
+            W = self._update_transforms(X, S, R, subjs)
+            S = self._update_outliers(X, W, R, subjs, self.lam)
+            R = self._update_shared_response(X, S, W, subjs, features, TRs)
+            # Print objective function every iteration
+            if logger.isEnabledFor(logging.INFO):
+                objective = self._objective_function(X, W, R, S, subjs,
+                                                     self.lam)
+                logger.info('Objective function %f' % objective)
+
+        return W, R, S
+
+    def _init_transforms(self, subjs, voxels, features, random_state):
+        """Initialize the mappings (Wi) with random orthogonal matrices.
+        Parameters
+        ----------
+
+        subjs : int
+            The number of subjects.
+
+        voxels : list of int
+            A list with the number of voxels per subject.
+
+        features : int
+            The number of features in the model.
+
+        random_state : `RandomState`
+            A random state to draw the mappings.
+
+        Returns
+        -------
+
+        W : list of array, element i has shape=[voxels_i, features]
+            The initialized orthogonal transforms (mappings) :math:`W_i` for
+            each subject.
+
+        Note
+        ----
+            Not thread safe.
+        """
+
+        # Init the Random seed generator
+        np.random.seed(self.rand_seed)
+
+        # Draw a random W for each subject
+        W = [random_state.random_sample((voxels[i], features))
+             for i in range(subjs)]
+        for i in range(subjs):
+            W[i], _ = np.linalg.qr(W[i])
+
+        return W
+
+    @staticmethod
+    def _objective_function(X, W, R, S, subjs, gamma):
+        """Evaluate the objective function.
+
+        .. math:: \sum_{i=1}^{N} 1/2 \| X_i - W_i R - S_i \|_F^2
+        .. math:: + /\gamma * \|S_i\|_1
+
+        Parameters
+        ----------
+
+        X : list of 2D arrays, element i has shape=[voxels_i, samples]
+            Each element in the list contains the fMRI data for alignment of
+            one subject.
+
+        W : list of array, element i has shape=[voxels_i, features]
+            The orthogonal transforms (mappings) :math:`W_i` for each subject.
+
+        R : array, shape=[features, samples]
+            The shared response.
+
+        S : list of array, element i has shape=[voxels_i, samples]
+            The individual component :math:`S_i` for each subject.
+
+        subjs : int
+            The number of subjects.
+
+        gamma : float, default: 1.0
+            Regularization parameter for the sparseness of the individual
+            components.
+
+        Returns
+        -------
+
+        func : float
+            The RSRM objective function evaluated on the parameters to this
+            function.
+        """
+        func = .0
```

Comment:
Some more description of this objective function would be appreciated. For instance, it is testing the difference between the individual data and the fitted response. However, I am unsure why you subtract the S rather than add or why you take the absolute sum of S or why you half the total

Reply:
This is the objective function that defines the problem (eq 2 in paper, will add it to description of method). In fact, Si is adding to the shared response term (Wi*R),  but it is subtracting from the original data Xi. The idea is to make sure that there is data fidelity (Wi*R+Si is approx. Xi). In other words, we try to minimize the distance between Xi and (Wi*R+Si). 
We are not taking the absolute value of Si but we are computing the l1-norm, which is known to obtain sparse solutions. In other words, we are requiring each Si to be sparse.

All the description is in the paper and this is an auxiliary internal function (for logging).

Reply:
Fair enough